src/relay/transforms/defunctionalization.cc

```diff
@@ -0,0 +1,428 @@
+    } else {
+      auto typedata = mod->LookupTypeDef(gtv);
+      auto constructors = typedata->constructors;
+      constructors.push_back(c);
+      mod->UpdateTypeDef(gtv, TypeData(typedata->header, typedata->type_vars, constructors));
+    }
+  }
+  /*!
+   * \brief add a case to the apply function, creating the function if it does not exist
+   *
+   * \param apply_gv GlobalVar of the apply function
+   * \param ft is the type functions the apply function handles
+   * \param c constructor to add a case for
+   * \param expr calls this expr with the args to the apply_gv
+   * \param patterns PatterVars to match with the constructor, used for handling free vars in
+   * functions
+   */
+  void AddApplyCase(GlobalVar apply_gv, FuncType ft, Constructor c, const Expr& expr,
+                    const Array<Pattern> patterns) {
+    CHECK(c->inputs.size() == patterns.size())
+        << "constructor function and pattern vars have different sizes";
+    if (!mod->ContainGlobalVar(apply_gv->name_hint)) {
+      auto x = Var("x", TypeCall(c->belong_to, {}));
+      auto vars = Array<Var>({x});
+      auto args = Array<Expr>();
+      for (auto t : ft->arg_types) {
+        auto y = Var("y", t);
+        vars.push_back(y);
+        args.push_back(y);
+      }
+
+      auto clauses = Array<Clause>({Clause(PatternConstructor(c, patterns), Call(expr, args))});
+      auto body = Match(x, clauses);
+      auto f = Function(vars, body, ft->ret_type, {});
+
+      mod->Add(apply_gv, f);
+    } else {
+      auto f = Downcast<Function>(mod->Lookup(apply_gv));
+      auto body = f->body.as<MatchNode>();
+      CHECK(body) << "internal invariant broken; apply function body should be a match node";
+
+      auto clauses = body->clauses;
+      auto x = f->params[0];
+      auto args = Array<Expr>();
+      for (size_t i = 1; i < f->params.size(); i++) {
+        args.push_back(f->params[i]);
+      }
+      clauses.push_back(Clause(PatternConstructor(c, patterns), Call(expr, args)));
+
+      mod->Add(apply_gv, Function(f->params, Match(x, clauses), f->ret_type, f->type_params), true);
+    }
+  }
+
+  /*!
+   * \brief encode a global var with a specialized type with a datatype
+   */
+  Expr EncodeGlobalVar(const GlobalVar& gv, const FuncType& ft) {
+    auto map = gv_datatype_map[gv];
+    auto type_key = TypeToString(ft);
+    if (map.count(type_key) == 0) {
+      auto gtv = GetFuncEncode(ft);
+      auto c = Constructor(std::to_string(constructor_counter++), {}, gtv);
+      map[type_key] = c;
+      AddConstructor(gtv, c);
+      AddApplyCase(GetApplyFunction(ft), ft, c, gv, {});
+    }
+    return Call(map[type_key], {});
+  }
+
+  /*!
+   * \brief type to string
+   */
+  std::string TypeToString(const Type& t) {
+    std::ostringstream s;
+    s << t;
+    return s.str();
+  }
+
+  /*!
+   * \brief get ADT handle for encoding type t
+   */
+  GlobalTypeVar GetFuncEncode(const Type& t) {
+    auto adt_name = "T" + TypeToString(t);
```

Comment:
```suggestion
    auto adt_name = "Defunc" + TypeToString(t);
```
a longger name to avoid name clashing